library/x509_crt.c

```diff
@@ -745,6 +778,84 @@ static int x509_get_subject_alt_name(unsigned char **p,
     return 0;
 }
 
+/*
+ * AuthorityKeyIdentifier ::= SEQUENCE {
+ *        keyIdentifier [0] KeyIdentifier OPTIONAL,
+ *        authorityCertIssuer [1] GeneralNames OPTIONAL,
+ *        authorityCertSerialNumber [2] CertificateSerialNumber OPTIONAL }
+ *
+ *    KeyIdentifier ::= OCTET STRING
+ */
+static int x509_get_authority_key_id(unsigned char **p,
+                                     unsigned char *end,
+                                     mbedtls_x509_authority *authority_key_id)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t len = 0u;
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
```

Comment:
If this check gets a length that's outside of bounds, it's silently ignored.
Please add a test for that and also check, if 
```
    if (*p + len != end) {
        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS,
                                 MBEDTLS_ERR_ASN1_LENGTH_MISMATCH);
    }
```

Reply:
Fixed.

Also adapted returned errors to indicate extension parsing failure (`MBEDTLS_ERR_X509_INVALID_EXTENSIONS`).

Reply:
We have an application which uses tiny-curl, mbedtls for SSL connections.
On our ESP-IDF v5.0.1, SSL connections were fine and able to communicate with Azure IoT.
Recently, we started upgrading ESP-IDF to v5.1.2 with mbedtls v3.5.0, we observed an error for "AuthorityKeyIdentifier" as follows.

Curl Debug Callback()[0]: * Error reading private key (nil) - mbedTLS: (-0x2566) X509 - The extension tag or value is invalid : ASN1 - Actual length differs from expected length

What are the possible cases when length can go outside of bounds?

Reply:
@BhanuPrakash-P In principle this error indicates that you have some bad data inside an X.509 structure. But X.509 is complicated and I can't rule out a bug or limitation in Mbed TLS. If you want to have any chance of getting help, please post an issue with an actual X.509 certificate and a small amount of code (or maybe one of the [sample programs](https://github.com/Mbed-TLS/mbedtls/tree/mbedtls-3.6.0/programs/x509)) that reproduce the problem.